collections/nemo_nlp/nemo_nlp/data/datasets/utils.py

```diff
@@ -778,21 +787,210 @@ def process_dialogflow(
+            for intent_name in f.readlines():
+                intent_names[intent_name.strip()] = str(intent_count)
+                intent_count += 1
+    print(intent_names)
+
+    for i, utterance in enumerate(utterances[1:]):
+
+        if utterance[1] not in agreedallDict:
+            agreedallDict[utterance[0]] = utterance[1]
+
+        if utterance[1] not in intent_names:
+            intent_names[utterance[1]] = str(intent_count)
+            intent_count += 1
+
+    print('x eval mechanism:')
+    print(len(agreedallDict))
+
+    return agreedallDict, intent_names
+
+
+def get_slot_labels(slot_annotations, task_name):
+
+    slot_labels = json.loads(slot_annotations[0])
+
+    all_labels = {}
+    count = 0
+    # Generating labels with the IOB format.
+    for label in slot_labels[task_name]['annotations']['labels']:
+        b_slot = ''.join(["B-", label['label']])
+        i_slot = ''.join(["I-", label['label']])
+        all_labels[b_slot] = str(count)
+        count += 1
+        all_labels[i_slot] = str(count)
+        count += 1
+    all_labels['O'] = str(count)
+
+    return all_labels
+
+
+def process_intent_slot_mturk(slot_annotations, agreedallDict, intent_names,
+                              task_name):
+
+    slot_tags = []
+    inorder_utterances = []
+    all_labels = get_slot_labels(slot_annotations, task_name)
+    print(f'agreedallDict - {len(agreedallDict)}')
+    print(f'Slot annotations - {len(slot_annotations)}')
+
+    for annotation in slot_annotations[0:]:
+        an = json.loads(annotation)
+        utterance = an['source']
+        if utterance.startswith('"') and utterance.endswith('"'):
+            utterance = utterance[1:-1]
+
+        if utterance in agreedallDict:
+            entities = {}
+            annotated_entities = an[task_name]['annotations']['entities']
+            for i, each_anno in enumerate(annotated_entities):
+                entities[int(each_anno['startOffset'])] = i
+
+            lastptr = 0
+            slots = ""
+            # sorting annotations by the start offset
+            for i in sorted(entities.keys()):
+                annotated_entities = an[task_name]['annotations']['entities']
+                tags = annotated_entities[entities.get(i)]
+                untagged_words = utterance[lastptr:tags['startOffset']]
+                for word in untagged_words.split():
+                    slots = ' '.join([slots, all_labels.get('O')])
+                anno_words = utterance[tags['startOffset']:tags['endOffset']]
+                # tagging with the IOB format.
+                for i, word in enumerate(anno_words.split()):
+                    if i == 0:
+                        b_slot = ''.join(["B-", tags['label']])
+                        slots = ' '.join([slots, all_labels.get(b_slot)])
+                    else:
+                        i_slot = ''.join(["I-", tags['label']])
+                        slots = ' '.join([slots, all_labels.get(i_slot)])
```

Comment:
strings are immutable so every time you call: slots = ' '.join, the old string is destroyed and a new string is created, which might lead to a lot of overhead for large data.
you might want to use slots as a list. so everytime you get a new slot, you append it to slots.
by the end, simple do a ' '.join(slots).